### ISSUE TITLE
`simpl` fails on mutual fixpoint with a section variable.

### ISSUE THREAD

**Issue by arthuraa** [coq/coq#12521]
#### Description of the problem

```coq
Inductive foo :=
| F0 : nat -> foo
| F1 : bar -> foo

with bar :=
| B1 : foo -> bar.

Section Def.

(* Putting this variable here breaks simpl below *)

Variable f : nat -> nat.

Fixpoint eval_foo x :=
  match x with
  | F0 n => f n
  | F1 x => eval_bar x
  end

with eval_bar x :=
  match x with
  | B1 x => eval_foo x
  end.

End Def.

Goal forall f x, eval_foo f (F1 x) = eval_bar f x.
Proof. simpl. (* Does *not* simplify the LHS *) Abort.

(* If we declare the variable for each function, the problem goes away. *)

Fixpoint eval_foo' f x : nat :=
  match x with
  | F0 n => f n
  | F1 x => eval_bar' f x
  end

with eval_bar' f x :=
  match x with
  | B1 x => eval_foo' f x
  end.

Goal forall f x, eval_foo' f (F1 x) = eval_bar' f x.
Proof. simpl. (* Does simplify the LHS *) reflexivity. Qed.

```

#### Coq Version

8.11.0

**Comment by Blaisorblade** [comment]
Likely duplicate of https://github.com/coq/coq/issues/3488.

**Comment by herbelin** [comment]
Fixed in PR #17993. Regarding #3488, this means that only the `simpl` part of #3488 is covered.
